### DISCUSSION TITLE
SSH Tunneling

### DISCUSSION THREAD

**Discussion by volga629-1** [kndndrj/nvim-dbee#56] [Ideas]
Hello Everyone,
Idea is add ability to define SSH tunnel  for database connectivity  based on ~/ssh/config or /etc/ssh (linux|mac)
Some  time direct ports not available and require ssh.

**Comment by volga629-1** [comment]
Example of  ssh jump 


```
Host dev-jump
    User 
    HostName db01
    IdentityFile  ~/.ssh/id_rsa
    ForwardAgent yes
```

**Comment by kndndrj** [comment]
I agree, built in ssh tunelling would be useful and really nice.

I'm not entirely sure what do you mean with the .ssh config file though. If I understand correctly, yoj want to specify a tunnel in the config file and then dbee would fire up the tunnel based on that connection?

How do you see the interface in dbee to look like?

Would something like this work for you:

```
postgres://connectionstring....?sshtunnel=dev-jump
```

I also don't know a lot about using ssh with go, so I'm not sure how easy it would be to parse the ssh config file (I don't think implementing our own parser is a good idea)

**Comment by volga629-1** [comment]
I think stage one is will be enough to add argument as you mention or in database client definition ssh: dev-jump ( do search based Host dev-jump in ~/.ssh/config

Example 

```
Host dev-jump
    User dev1
    HostName dev-jump
    IdentityFile  ~/.ssh/id_rsa
    ForwardAgent yes

Host dev-db-fwd
  User dev
  Hostname dev-dev-db-fwd
  IdentityFile ~/.ssh/key_dev_db
  Localforward 5470 pgsql:5432
  Proxyjump dev-jump
```

New Connection string 

```
name: my db
ssh:   dev-db-fwd
type: postgres
url: postgres://username:password@127.0.0.1:5470/mydatabse?sslmode=verify
page size: 500

```

**Comment by volga629-1** [comment]
Go ssh 

```
https://pkg.go.dev/golang.org/x/crypto/ssh
```

**Comment by kndndrj** [comment]
I guess it would make sense to look how other database clients do these things as well

**Reply by volga629-1** [reply]
That will good idea too

**Comment by TelpeNight** [comment]
Hi. I'm also looking for ssh support.

I've developed https://github.com/TelpeNight/mytunnel for my work project. Using it every day. It would be trivial to add ssh support for mysql at least, if you'll adopt it.
